<main>
<article class="day-desc"><h2>--- Day 1: Trebuchet?! ---</h2><p>Something is wrong with global snow production, and you've been selected to take a look. The Elves have even given you a map; on it, they've used stars to mark the top fifty locations that are likely to be having problems.</p>
<p>You've been doing this long enough to know that to restore snow operations, you need to check all <em class="star">fifty stars</em> by December 25th.</p>
<p>Collect stars by solving puzzles.  Two puzzles will be made available on each day in the Advent calendar; the second puzzle is unlocked when you complete the first.  Each puzzle grants <em class="star">one star</em>. Good luck!</p>
<p>You try to ask why they can't just use a <a href="/2015/day/1">weather machine</a> ("not powerful enough") and where they're even sending you ("the sky") and why your map looks mostly blank ("you sure ask a lot of questions") <span title="My hope is that this abomination of a run-on sentence somehow conveys the chaos of being hastily loaded into a trebuchet.">and</span> hang on did you just say the sky ("of course, where do you think snow comes from") when you realize that the Elves are already loading you into a <a href="https://en.wikipedia.org/wiki/Trebuchet" target="_blank">trebuchet</a> ("please hold still, we need to strap you in").</p>
<p>As they're making the final adjustments, they discover that their calibration document (your puzzle input) has been <em>amended</em> by a very young Elf who was apparently just excited to show off her art skills. Consequently, the Elves are having trouble reading the values on the document.</p>
<p>The newly-improved calibration document consists of lines of text; each line originally contained a specific <em>calibration value</em> that the Elves now need to recover. On each line, the calibration value can be found by combining the <em>first digit</em> and the <em>last digit</em> (in that order) to form a single <em>two-digit number</em>.</p>
<p>For example:</p>
<pre><code>1abc2
pqr3stu8vwx
a1b2c3d4e5f
treb7uchet
</code></pre>
<p>In this example, the calibration values of these four lines are <code>12</code>, <code>38</code>, <code>15</code>, and <code>77</code>. Adding these together produces <code><em>142</em></code>.</p>
<p>Consider your entire calibration document. <em>What is the sum of all of the calibration values?</em></p>
</article>



1abc2=12
pqr3stu8vwx=38
a1b2c3d4e5f=15
treb7uchet=77

outer numbers and if in middle then inner

In [229]:
import pandas as pd
import re

In [230]:
data = pd.read_csv('input.txt', names=['A'])
data


A
0            mxmkjvgsdzfhseightonetwoeight7
1                   3five4s84four9rtbzllggz
2                      75sevenzdrpkv1onetwo
3                        3q7ctkghhqkpb5four
4    ccxpblrgrjxcgrhjxfmtwonine8eightzzrdn4
..                                      ...
995      cnktjkjmcg46fiverxlxkmxvkmnklsfive
996                         5lvlhsjkxssfour
997                   sevenfourfour99seven8
998                         ktgfiveone76ghj
999                    7zgzsevenftkdfour186

[1000 rows x 1 columns]

In [231]:
nums = data['A'].apply(lambda x: re.findall(r'\d',x))
nums

0                  [7]
1      [3, 4, 8, 4, 9]
2            [7, 5, 1]
3            [3, 7, 5]
4               [8, 4]
            ...       
995             [4, 6]
996                [5]
997          [9, 9, 8]
998             [7, 6]
999       [7, 1, 8, 6]
Name: A, Length: 1000, dtype: object

In [232]:
def solve(numsToSum): 
    numsToSum = [int(i) for i in numsToSum]
    if len(numsToSum) < 2:
        return numsToSum[0]*10 + numsToSum[0]
    else:
        return numsToSum[0]*10 + numsToSum[-1]

In [233]:
nums['test'] = nums.apply(solve)

In [234]:
nums['test']

0      77
1      39
2      71
3      35
4      84
       ..
995    46
996    55
997    98
998    76
999    76
Name: A, Length: 1000, dtype: int64

In [235]:
mySum = nums.test.sum()

In [236]:
mySum

55621

---

begin part 2



<p>Your puzzle answer was <code>55621</code>.</p><p class="day-success">The first half of this puzzle is complete! It provides one gold star: *</p>
<article class="day-desc"><h2 id="part2">--- Part Two ---</h2><p>Your calculation isn't quite right. It looks like some of the digits are actually <em>spelled out with letters</em>: <code>one</code>, <code>two</code>, <code>three</code>, <code>four</code>, <code>five</code>, <code>six</code>, <code>seven</code>, <code>eight</code>, and <code>nine</code> <em>also</em> count as valid "digits".</p>
<p>Equipped with this new information, you now need to find the real first and last digit on each line. For example:</p>
<pre><code>two1nine
eightwothree
abcone2threexyz
xtwone3four
4nineeightseven2
zoneight234
7pqrstsixteen
</code></pre>
<p>In this example, the calibration values are <code>29</code>, <code>83</code>, <code>13</code>, <code>24</code>, <code>42</code>, <code>14</code>, and <code>76</code>. Adding these together produces <code><em>281</em></code>.</p>
<p><em>What is the sum of all of the calibration values?</em></p>
</article>


In [237]:
data = pd.read_csv('input.txt', names=['A'])
data

A
0            mxmkjvgsdzfhseightonetwoeight7
1                   3five4s84four9rtbzllggz
2                      75sevenzdrpkv1onetwo
3                        3q7ctkghhqkpb5four
4    ccxpblrgrjxcgrhjxfmtwonine8eightzzrdn4
..                                      ...
995      cnktjkjmcg46fiverxlxkmxvkmnklsfive
996                         5lvlhsjkxssfour
997                   sevenfourfour99seven8
998                         ktgfiveone76ghj
999                    7zgzsevenftkdfour186

[1000 rows x 1 columns]

In [238]:
numberRegex = r'(?=(one|two|three|four|five|six|seven|eight|nine|zero|\d))'
nums['A'] = data['A'].apply(lambda x: re.findall(numberRegex,x))
nums

0                                                     [7]
1                                         [3, 4, 8, 4, 9]
2                                               [7, 5, 1]
3                                               [3, 7, 5]
4                                                  [8, 4]
                              ...                        
997                                             [9, 9, 8]
998                                                [7, 6]
999                                          [7, 1, 8, 6]
test    0      77
1      39
2      71
3      35
4     ...
A       0              [eight, one, two, eight, 7]
1  ...
Name: A, Length: 1002, dtype: object

In [239]:
m = {
    "one":1,
    "two":2,
    "three":3,
    "four":4,
    "five":5,
    "six":6,
    "seven":7,
    "eight":8,
    "nine":9,
    "zero":0
}

nums['A']=nums['A'].apply(lambda x: [m.get(y, y) for y in x] )



In [240]:
nums['test'] = nums.A.apply(solve)


In [241]:
nums['test']

0      87
1      39
2      72
3      34
4      24
       ..
995    45
996    54
997    78
998    56
999    76
Name: A, Length: 1000, dtype: int64

In [242]:
nums['A']

0            [8, 1, 2, 8, 7]
1      [3, 5, 4, 8, 4, 4, 9]
2         [7, 5, 7, 1, 1, 2]
3               [3, 7, 5, 4]
4            [2, 9, 8, 8, 4]
               ...          
995             [4, 6, 5, 5]
996                   [5, 4]
997    [7, 4, 4, 9, 9, 7, 8]
998             [5, 1, 7, 6]
999       [7, 7, 4, 1, 8, 6]
Name: A, Length: 1000, dtype: object

In [243]:
part2Sum = nums['test'].sum()
part2Sum

53592